File: oolab/src/main/java/agh/ics/oop/model/Animal.java
package agh.ics.oop.model;

public class Animal implements WorldElement {
    private MapDirection currentOrientation;
    private Vector2d currentPosition;
    public static final Vector2d UPPER_RIGHT_LIMIT = new Vector2d(4, 4);
    public static final Vector2d LOWER_LEFT_LIMIT = new Vector2d(0, 0);
    public Animal(){
        this.currentOrientation = MapDirection.NORTH;
        this.currentPosition = new Vector2d(2, 2);


    }
    public Animal(Vector2d newPosition){
        this.currentOrientation = MapDirection.NORTH;
        this.currentPosition = newPosition;
    }

    public String orientationToString(){
        return switch (this.currentOrientation){
            case NORTH -> "^";
            case SOUTH -> "v";
            case EAST -> ">";
            case WEST -> "<";
        };

    }
    @Override
    public String toString(){
        return orientationToString();
    }

    @Override
    public boolean isAt(Vector2d position){
        return currentPosition.equals(position);
    }

    public boolean canMoveTo(Vector2d position){
        return LOWER_LEFT_LIMIT.precedes(position) && UPPER_RIGHT_LIMIT.follows(position);
    }

    public Vector2d calculateNextPosition(MoveDirection direction){
        return switch (direction){
            case FORWARD-> currentPosition.add(currentOrientation.toUnitVector());
            case BACKWARD -> currentPosition.subtract(currentOrientation.toUnitVector());
            default -> currentPosition;

        };
    }
    public void move(MoveDirection direction, MoveValidator validator){
        Vector2d newPosition = calculateNextPosition(direction);
        if(validator.canMoveTo(newPosition)){
            switch (direction){
                case RIGHT -> currentOrientation = currentOrientation.next();
                case LEFT -> currentOrientation = currentOrientation.previous();
                case FORWARD, BACKWARD -> currentPosition = newPosition;
            }
        }


    }
    @Override
    public MapDirection getCurrentOrientation() {
        return currentOrientation;
    }
    @Override
    public Vector2d getCurrentPosition() {
        return currentPosition;
    }


}

File: kotlin_lab/src/main/kotlin/Animal.kt
class Animal()  {
    private var currentOrientation : MapDirection = MapDirection.NORTH
    internal var currentPosition: Vector2d =  Vector2d(2,2)
        set(value) {
            field = value
        }


    private val upperRightLimit:Vector2d = Vector2d(4, 4)
    private val lowerLeftLimit:Vector2d = Vector2d(0,0)

    constructor(position: Vector2d) : this() {
        this.currentOrientation =MapDirection.NORTH
        this.currentPosition =position
    }


    private fun orientationToString() : String =
        when (this.currentOrientation){
            MapDirection.NORTH -> "^"
            MapDirection.SOUTH -> "v"
            MapDirection.EAST -> ">"
            MapDirection.WEST -> "<"
        }

    override  fun toString():String = orientationToString()

    fun isAt(position: Vector2d) :Boolean = currentPosition == position


    private fun canMoveTo(position: Vector2d):Boolean = (lowerLeftLimit <= position && position <= upperRightLimit)

private fun calculateNextPosition(direction: MoveDirection): Vector2d {
    val unitVector = currentOrientation.toUnitVector()
    return when (direction) {
        MoveDirection.FORWARD -> currentPosition + unitVector
        MoveDirection.BACKWARD -> currentPosition - unitVector
        else -> currentPosition
    }
}
    private fun  MapDirection.toUnitVector(): Vector2d = when (this) {
        MapDirection.NORTH -> Vector2d(0,1)
        MapDirection.SOUTH -> Vector2d(0, -1)
        MapDirection.EAST -> Vector2d(1, 0)
        MapDirection.WEST -> Vector2d(-1, 0)
    }



    fun move(direction: MoveDirection){
        val newPosition: Vector2d = calculateNextPosition(direction)
        if(canMoveTo(newPosition)){
            when(direction){
                MoveDirection.RIGHT -> currentOrientation = currentOrientation.next()
                MoveDirection.LEFT -> currentOrientation = currentOrientation.previous()
                MoveDirection.BACKWARD, MoveDirection.FORWARD -> currentPosition = newPosition
            }
        }

    }



    fun getCurrentOrientation(): MapDirection = currentOrientation

    fun getCurrentPosition():Vector2d = currentPosition








}

